src/lib/zf/zf_ds.c

```diff
@@ -166,10 +166,11 @@ int zf_delegated_send_complete(struct zft* ts, const struct iovec* iov,
     already_acked = 0;
 
     /* Queue remaining packets on send queue */
+    uint32_t current_lbb = pcb->snd_lbb;
     rc = tcp_queue_sent_segments(tcp, &pcb->sendq, &iov_temp, &pcb->snd_lbb);
     if( ZF_UNLIKELY(rc < 0) )
       goto out;
-    tcp_output_timers_common(zf_stack_from_zocket(tcp), tcp, pcb->snd_lbb);
+    tcp_output_timers_common(zf_stack_from_zocket(tcp), tcp, current_lbb);
```

Comment:
I did wonder if this be the below instead?
```suggestion
    tcp_output_timers_common(zf_stack_from_zocket(tcp), tcp, pcb->snd_lbb - 1);
```

Based on my reading of the delegated sends API (from [trader_tcpdirect_ds_efvi](https://github.com/Xilinx-CNS/tcpdirect/blob/master/src/tests/trade_sim/trader_tcpdirect_ds_efvi.c)), `zf_delegated_send_complete` is called to tell TCPDirect that we have sent a packet and it should update its internal state accordingly (e.g., [note this being called after an ef_vi_transmit in the test app](https://github.com/Xilinx-CNS/tcpdirect/blob/14d93228f700fad32ea7fa93f88d79d59d7748ad/src/tests/trade_sim/trader_tcpdirect_ds_efvi.c#L279)). In your example, sending sequences [B, E) you note that `tcp_queue_sent_segments` will update `pcb->snd_lbb` to E (i.e., the next sequence number to be sent, or equivalently, one plus the last sequence number sent). In which case I thought it might make sense to wait for D to estimate the RTT.

However, I expect this suggestion could lead to overestimation of the RTT as it really measures the RTT of three packets (sent at the same time) in this example. In which case your code which only estimates the RTT for the first packet seems more reasonable. Does this make sense, or have I made a mistake somewhere along the way?

Reply:
Yes, I believe your understanding is correct. The intention is to estimate the current RTT as simple and consistent as possible. 

Regarding your suggestion above, I'd recommend to reference the first byte sent for the following reasons:

- TCP is a byte oriented protocol and there might be disagreements between the sender and the receiver in the MSS; see Section 4.2 of https://datatracker.ietf.org/doc/html/rfc2581

- See also Karn and Partridge, "Improving Round-Trip Time Estimates in Reliable Transport Protocols" for further complications - the takeaway of this paper is to keep the current RTT estimation as simple and as consistent as possible. I mean, consistent with other parts of the code both in TCP Direct and in Onload - both products time the first byte of the most recent segment  except for the bug in question. 

- I also believe there could be some sort of head of line blocking for estimating the current RTT (if we adopted `pcb->snd_lbb-1`). So, apart from the original (most recent) segment, which initial byte is `pcb->snd_lbb` (before it is modified), there may be other segments sent along that one (via iov). By timing the latest byte (`pcb->snd_lbb-1`), we are decreasing the chances of obtaining the sample for the most recent segment (`pcb->snd_lbb` before it is modified) as we do not know the fate for accompanying segments (e.g., those can be reordered or dropped) thereby impairing the current RTT estimation.

Reply:
I wonder if the viable/correct minimal change is to move the `tcp_output_timers_common` call before sending the sent segments to the queue. I don't see any problem yet, and it does work in my tests. 

```suggestion
    tcp_output_timers_common(zf_stack_from_zocket(tcp), tcp, pcv->snd_lbb);
    rc = tcp_queue_sent_segments(tcp, &pcb->sendq, &iov_temp, &pcb->snd_lbb);
    if( ZF_UNLIKELY(rc < 0) )
      goto out;
```

A potential problem could be related to `rc < 0`, however I believe it happens when the sent queue is too long and send needs to be retried, in which case `tcp_output_timers_common` is going to be called again.

Reply:
Thanks for the explanation and resources, they were quite helpful to consolidate my understanding!

Based on what you've said so far I do think the change you suggest in the above comment might be what we want. Notably, we set `pcb->rttest` to the current tick in `tcp_ouput_timers_common`, which is [later used to calculate the sample RTT](https://github.com/Xilinx-CNS/tcpdirect/blob/14d93228f700fad32ea7fa93f88d79d59d7748ad/src/include/zf_internal/tcp_rx.h#L62). With the current code, an app might fail here, poll the reactor, and try calling `zf_delegated_send_complete` again, which would have a different (later) timer tick than your new suggestion. In the newly suggested code, we would call it multiple times, but this would only do work the first time when `pcb->rttest` is 0.

Reply:
I'd also add that with the new suggestion the `ZF_TCP_TIMER_RTO` would start at the right time (if need be) if `tcp_queue_sent_segments` fails.